contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaRecordReader.java

```diff
@@ -0,0 +1,178 @@
+    Set<SchemaPath> transformed = Sets.newLinkedHashSet();
+    if (!isStarQuery()) {
+      for (SchemaPath column : projectedColumns) {
+        transformed.add(column);
+      }
+    } else {
+      transformed.add(Utilities.STAR_COLUMN);
+    }
+    return transformed;
+  }
+
+  @Override
+  public void setup(OperatorContext context, OutputMutator output) throws ExecutionSetupException {
+    this.writer = new VectorContainerWriter(output, unionEnabled);
+    messageReader = MessageReaderFactory.getMessageReader(plugin.getConfig().getDrillKafkaProps());
+    messageReader.init(context.getManagedBuffer(), Lists.newArrayList(getColumns()), this.writer,
+        this.enableAllTextMode, false, this.readNumbersAsDouble);
+    kafkaConsumer = messageReader.getConsumer(plugin);
+    partitions = Lists.newArrayListWithCapacity(1);
+    TopicPartition topicPartition = new TopicPartition(subScanSpec.getTopicName(), subScanSpec.getPartitionId());
+    partitions.add(topicPartition);
+    kafkaConsumer.assign(partitions);
+    currentOffset = subScanSpec.getStartOffset();
+    kafkaConsumer.seek(topicPartition, currentOffset);
+  }
+
+  private boolean hasNext() {
+    if (currentOffset > endOffset) {
+      logger.info("Read all the messages from Kafka. Start offset : {}, End offset : {}", currentOffset, endOffset);
+      return false;
+    } else if (messageIter != null && messageIter.hasNext()) {
+      return true;
+    } else {
+      return fetch();
+    }
+  }
+
+  private boolean fetch() {
+    long startTime = System.currentTimeMillis();
+    ConsumerRecords<byte[], byte[]> consumerRecords = kafkaConsumer.poll(kafkaPollTimeOut);
+
+    if (consumerRecords.isEmpty()) {
+      logger.warn("Not able to fetch messages within {}milliseconds. Consider increasing the value of {}",
+          DRILL_KAFKA_POLL_TIMEOUT, kafkaPollTimeOut);
+    }
+
+    long lastFetchTime = (System.currentTimeMillis() - startTime);
+    logger.debug("Total number of messages fetched : " + consumerRecords.count());
+    logger.debug("Time taken to fetch : " + (lastFetchTime / 1000) + " seconds");
+    totalFetchTime += lastFetchTime;
+
+    messageIter = consumerRecords.iterator();
+    return messageIter.hasNext();
+  }
+
+  /**
+   * KafkaConsumer.poll will fetch 500 messages per poll call. So hasNext will
+   * take care of polling multiple times for this given batch next invocation
+   */
+  @Override
+  public int next() {
+    writer.allocate();
+    writer.reset();
+    Stopwatch watch = Stopwatch.createStarted();
+
+    int messageCount = 0;
+    while (hasNext()) {
+      ConsumerRecord<byte[], byte[]> consumerRecord = messageIter.next();
+      currentOffset = consumerRecord.offset();
+      writer.setPosition(messageCount);
+      messageReader.readMessage(consumerRecord);
+      ++messageCount;
+      if (messageCount >= DEFAULT_MESSAGES_PER_BATCH) {
+        break;
+      }
```

Comment:
Seems awkward. Do you get to control the number of messages per fetch? If so, can you loop over chunks to fill the batch, avoiding the need for chunks to span batches?
A sketch:
```
while (messageCount < DEFAULT_MESSAGES_PER_BATCH) {
  int fetchcount = Math.min(500, DEFAULT_MESSAGES_PER_BATCH - messageCount);
  messageIter = readNextChunk();
  for (ConsumerRecord<byte[], byte[]> consumerRecord : messageIter) {
    // Process record
    messageCount++;
  }
}
```

That said, however, our eventual goal is to limit batch sizes based on memory use, not just row count. At that point, the current design will work better as you'll need to carry over the iterator from one batch to the next because you won't know, in general, how many records a batch can contain.